ml-agents/mlagents/trainers/components/reward_signals/gail/model.py

```diff
@@ -0,0 +1,261 @@
+
+        self.encoded_expert = tf.concat(encoded_expert_list, axis=1)
+        self.encoded_policy = tf.concat(encoded_policy_list, axis=1)
+
+    def create_encoder(
+        self, state_in: tf.Tensor, action_in: tf.Tensor, done_in: tf.Tensor, reuse: bool
+    ) -> Tuple[tf.Tensor, tf.Tensor]:
+        """
+        Creates the encoder for the discriminator
+        :param state_in: The encoded observation input
+        :param action_in: The action input
+        :param done_in: The done flags input
+        :param reuse: If true, the weights will be shared with the previous encoder created
+        """
+        with tf.variable_scope("GAIL_model"):
+            if self.use_actions:
+                concat_input = tf.concat([state_in, action_in, done_in], axis=1)
+            else:
+                concat_input = state_in
+
+            hidden_1 = tf.layers.dense(
+                concat_input,
+                self.h_size,
+                activation=tf.nn.elu,
+                name="d_hidden_1",
+                reuse=reuse,
+            )
+
+            hidden_2 = tf.layers.dense(
+                hidden_1,
+                self.h_size,
+                activation=tf.nn.elu,
+                name="d_hidden_2",
+                reuse=reuse,
+            )
+
+            z_mean = None
+            if self.use_vail:
+                # Latent representation
+                z_mean = tf.layers.dense(
+                    hidden_2,
+                    self.z_size,
+                    reuse=reuse,
+                    name="z_mean",
+                    kernel_initializer=LearningModel.scaled_init(0.01),
+                )
+
+                self.noise = tf.random_normal(tf.shape(z_mean), dtype=tf.float32)
+
+                # Sampled latent code
+                self.z = z_mean + self.z_sigma * self.noise * self.use_noise
+                estimate_input = self.z
+            else:
+                estimate_input = hidden_2
+
+            estimate = tf.layers.dense(
+                estimate_input,
+                1,
+                activation=tf.nn.sigmoid,
+                name="d_estimate",
+                reuse=reuse,
+            )
+            return estimate, z_mean
+
+    def create_network(self) -> None:
+        """
+        Helper for creating the intrinsic reward nodes
+        """
+        if self.use_vail:
+            self.z_sigma = tf.get_variable(
+                "sigma_vail",
+                self.z_size,
+                dtype=tf.float32,
+                initializer=tf.ones_initializer(),
+            )
+            self.z_sigma_sq = self.z_sigma * self.z_sigma
+            self.z_log_sigma_sq = tf.log(self.z_sigma_sq + 1e-7)
+            self.use_noise = tf.placeholder(
+                shape=[1], dtype=tf.float32, name="NoiseLevel"
+            )
+        self.expert_estimate, self.z_mean_expert = self.create_encoder(
+            self.encoded_expert, self.expert_action, self.done_expert, False
```

Comment:
style nit: can you do `reuse=False` (and `reuse=True`) instead? I know it's not strictly necessary, but (imho) makes it clearer when reading (without having to look at the definition of create_encoder).